heirloom/diff/diff.1
.\"
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	from 4.3BSD diff.1	6.4 (Berkeley) 5/19/86
.\"
.\" This code contains changes by
.\" Gunnar Ritter, Freiburg i. Br., Germany, March 2003. All rights reserved.
.\"
.\" Conditions 1, 2, and 4 and the no-warranty notice below apply
.\" to these changes.
.\"
.\"
.\" Copyright (c) 1980, 1993
.\" 	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowedgement:
.\" 	This product includes software developed by the University of
.\" 	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.TH DIFF 1 "3/26/05" "Heirloom Toolchest" "User Commands"
.SH NAME
diff \- differential file comparator
.SH SYNOPSIS
.HP
.nh
.ad l
\fBdiff\fR
[\fB\-abBiptw\fR]
[\fB\-cefhnu\fR]
[\fB\-C\ \fInumber\fR]
[\fB\-U\ \fInumber\fR]
\fIfile1\fR \fIfile2\fR
.HP
.nh
.ad l
\fBdiff\fR
[\fB\-abBiptw\fR]
[\fB\-D\ \fIstring\fR]
\fIfile1\fR \fIfile2\fR
.HP
.nh
.ad l
\fBdiff\fR
[\fB\-abBiNptw12\fR]
[\fB\-lrs\fR]
[\fB\-cefhnu\fR]
[\fB\-C\ \fInumber\fR]
[\fB\-S\ \fIname\fR]
[\fB\-U\ \fInumber\fR]
[\fB\-x\ \fIpattern\fR]
\fIdirectory1\fR \fIdirectory2\fR
.br
.hy 1
.SH DESCRIPTION
.I Diff
tells what lines must be changed in two files to bring them
into agreement.
If
.I file1
.RI ( file2 )
is `\-', the standard input is used.
If
.I file1
.RI ( file2 )
is a directory, then a file in that directory
whose file-name is the same as the file-name of
.I file2
.RI ( file1 )
is used.
The normal output contains lines of these forms:
.IP "" 5
.I n1
a
.I n3,n4
.br
.I n1,n2
d
.I n3
.br
.I n1,n2
c
.I n3,n4
.PP
These lines resemble
.I ed
commands to convert
.I file1
into
.IR file2 .
The numbers after the letters pertain to
.IR file2 .
In fact, by exchanging `a' for `d' and reading backward
one may ascertain equally how to convert 
.I file2
into
.IR file1 .
As in 
.I ed,
identical pairs where
.I n1
=
.I n2
or
.I n3
=
.I n4
are abbreviated as a single number.
.PP
Following each of these lines come all the lines that are
affected in the first file flagged by `<', 
then all the lines that are affected in the second file
flagged by `>'.
.TP 10
.B \-a
causes a list of differences to be output
for all files,
even for those found to have binary content.
This option is an extension.
.TP 10
.B \-b
causes trailing whitespace characters
to be ignored, and other
strings of whitespace to compare equal.
.TP 10
.B \-i
ignores the case of letters.  E.g., ``A'' will compare equal to ``a''.
.TP 10
.B \-p
causes the name of the surrounding C function,
or, more exactly,
of the first previous unchanged line
beginning with a letter, the dollar sign, or the underscore,
to be output with each set of changes.
Implies
.I \-c
unless
.I \-u
is also present.
This option is an extension.
.TP 10
.B \-t
will expand tabs in output lines.  Normal,
.B \-c
or
.B \-u
output adds character(s) to the front of each line which may screw up
the indentation of the original source lines and make the output listing
difficult to interpret.  This option will preserve the original source's
indentation.
.TP 10
.B \-w
is similar to
.B \-b
but causes whitespace characters
to be totally ignored.
E.g., ``if\ (\ a\ ==\ b\ )'' will compare equal to ``if(a==b)''.
.TP 10
.B \-B
causes changes that consist entirely of empty lines added or deleted
to be ignored.
This option is an extension.
.PP
The following options are mutually exclusive:
.TP 10
.B \-c
produces a diff with three lines of context.
With
.B \-c
the output format is modified slightly:
the output beginning with identification of the files involved and
their creation dates and then each change is separated
by a line with a dozen *'s.
The lines removed from
.I file1
are marked with `\(mi '; those added to
.I file2
are marked `+ '.  Lines which are changed from one
file to the other are marked in both files with with `! '.
.\".sp
.\"Changes which lie within <context> lines of each other are grouped
.\"together on output.  (This is a change from the previous ``diff -c''
.\"but the resulting output is usually much easier to interpret.)
.TP 10
\fB\-C\ \fInumber\fR
Same as
.B \-c
but uses
.I number
of lines of context.
.TP 10
\fB\-D\ \fIstring\fR
causes
.I diff
to create a merged version of
.I file1
and
.I file2
on the standard output, with C preprocessor controls included so that
a compilation of the result without defining \fIstring\fR is equivalent
to compiling
.I file1,
while defining
.I string
will yield
.I file2.
.TP 10
.B \-e
produces a script of
.I "a, c"
and 
.I d
commands for the editor
.I ed,
which will recreate
.I file2
from
.IR file1 .
In connection with
.BR \-e ,
the following shell program may help maintain
multiple versions of a file.
Only an ancestral file ($1) and a chain of 
version-to-version
.I ed
scripts ($2,$3,...) made by
.I diff
need be on hand.
A `latest version' appears on
the standard output.
.IP
\ \ \ \ \ \ \ \ (shift; cat $*; echo \'1,$p\') \(bv ed \- $1
.IP
Extra commands are added to the output when comparing directories with
.B \-e,
so that the result is a
.IR sh (1)
script for converting text files which are common to the two directories
from their state in
.I dir1
to their state in
.I dir2.
.TP 10
.B \-f
produces a script similar to that of
.B \-e,
not useful with
.I ed,
and in the opposite order.
.TP 10
.B \-h
does a fast, half-hearted job.
It works only when changed stretches are short
and well separated,
but does work on files of unlimited length.
.TP 10
.B \-n
produces a script similar to that of
.B \-e,
but in the opposite order and with a count of changed lines on each
insert or delete command.
.\"This is the form used by
.\".IR rcsdiff (1).
.TP 10
.B \-u
produces a unified diff with three lines of context.
The output begins with identification of the files involved
and their creation dates,
followed by the changes
separated by `@@ \-range +range @@'.
Lines removed from
.I file1
are marked with `\(mi',
those added to
.I file2
are marked `+'.
This option is an extension.
.TP 10
\fB\-U\ \fInumber\fR
Same as
.B \-u
but uses
.I number
of lines of context.
This option is an extension.
.PP
If both arguments are directories,
.I diff
sorts the contents of the directories by name, and then runs the
regular file
.I diff
algorithm on text files which are different.
Binary files which differ,
common subdirectories, and files which appear in only one directory
are listed.
.PP
Options when comparing directories are:
.TP 10
.B \-l
long output format; each text file
.I diff
is piped through
.IR pr (1)
to paginate it,
other differences are remembered and summarized
after all text file differences are reported.
.TP 10
.B \-N
causes the text of files
that exist in one directory only
to be output
as if compared to an empty file modified at 1/1/70.
This option is an extension.
.TP 10
.B \-1
is similar to
.IR \-N ,
but causes just the text of files that exist in
.I directory1
only to be output.
Files that exist only in
.I directory2
are listed.
This option is an extension.
.TP 10
.B \-2
is similar to
.IR \-N ,
but causes just the text of files that exist in
.I directory2
only to be output.
Files that exist only in
.I directory1
are listed.
This option is an extension.
.TP 10
.B \-r
causes application of
.I diff
recursively to common subdirectories encountered.
.TP 10
.B \-s
causes 
.I diff
to report files which are the same, which are otherwise not mentioned.
.TP 10
.BI \-S " name"
starts a directory
.I diff
in the middle beginning with file
.I name.
.TP 10
.BI \-x " pattern"
excludes all file names that match
.I pattern
(as described in
.IR glob (7))
from comparison.
If
.I pattern
matches a directory,
files below that directory are also excluded.
This option is an extension.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale(7).
.TP
.B LC_CTYPE
Sets the mapping of bytes to characters,
character case translation
and the set of whitespace characters.
.TP
.B SYSV3
If this variable is set,
invalid options are ignored instead of being rejected,
and the text of some diagnostic messages is changed.
.SH FILES
.ta \w'/usr/5lib/diffh   'u
/tmp/d?????
.br
/usr/5lib/diffh	for \fB\-h\fR
.br
diff	for directory diffs
.br
pr
.SH "SEE ALSO"
cmp(1),
cc(1),
comm(1),
ed(1),
diff3(1),
patch(1),
locale(7)
.SH DIAGNOSTICS
Exit status is 0 for no differences, 1 for some, 2 for trouble.
.SH NOTES
Editing scripts produced under the
.BR \-e " or"
.BR \-f " option are naive about"
creating lines consisting of a single `\fB.\fR'.
.PP
When comparing directories with the
\fB\-b, -w\fP or \fB-i\fP
options specified,
.I diff
first compares the files ala
.I cmp,
and then decides to run the
.I diff
algorithm if they are not equal.
This may cause a small amount of spurious output if the files
then turn out to be identical because the only differences are
insignificant blank string or case differences.
.PP
When
.I diff
output is used with
.IR ed (1)
or
.IR patch (1)
for file synchronization,
it is recommended that it is run in the
.I C
or another single-byte LC_CTYPE locale
since character-to-byte conversion
might otherwise hide some changes.


ex-vi/Changes
Release ...
* When a multicolumn character was replaced by another multicolumn character
  in insert mode, the display was not updated appropriately with terminals
  other than xterm.
* Various rendering errors happened with multicolumn characters if they
  started at an even column (counting from 1 upwards).
* When a multicolumn character was inserted and then replaced, the visual
  screen representation was sometimes not updated accordingly.
* Undoing the replacement of a multicolumn character by a singlecolumn
  character twice made the singlecolumn character invisible.
* The 'cw' visual command with a multibyte character as last character of
  the affected word located at the end of the line left garbage bytes past
  the end of the new text.
* Visual 'U' followed by 'u' lead to garbage on the screen when multibyte
  characters were present on the changed line.
* The handling of multicolumn characters was fixed for terminals without the
  IM (insert mode) capability. It is unlikely that such terminals actually
  exist, but vi will use the corresponding code for open mode if a termcap
  entry is not available.
* When an illegal multibyte sequence is entered in vi insert mode, no garbage
  bytes are generated anymore when the insert mode is left.
* The '\u', '\l', '\U', and '\L' substitution sequences work with multibyte
  characters now.
* Handle character case conversions with the '~' vi command correctly if the
  length of the converted multibyte sequence is smaller than the original one.
* Multibyte sequences that correspond to an unprintable character are now
  printed as multiple octal escape sequences.
* Fixed an old vi bug: If a vi command that yanked or deleted part of a line
  was followed by an ex command that also yanked or deleted some text, a
  following 'p' vi command pasted the text affected by the former vi command.
  It now pastes the text of the last yank or delete even if that was an ex
  command.

Release 12/2/04
* Support for multibyte character locales was added.
* The code has been converted to ANSI C, and support for pre-POSIX systems has
  been dropped.
* When the end of the current line consists of blank characters and the 'w'
  visual command is given at this point, vi now advances to the start of the
  next line instead of ringing the bell. This is compatible with SVR4 vi and
  seems to be what POSIX specifies.
* If the replacement part of a substitute command consists of a single '%',
  as in ':s/foo/%/', the replacement part of the previous substitution is
  used. This is compatible with SVR4 vi and is specified by POSIX.
* Fixed a number of possible heap overflows, e.g. because of too long tag
  strings.

Release 6/5/04
* Some changes were made to the Makefile to support RPM builds. In particular,
  the meaning of the DESTDIR and PREFIX variables was changed.
* An insufficient size of a variable caused the window size set to 8 on
  FreeBSD if the terminal baud rate was 38400.

Release 1/3/04
* Changes to the included libterm only: Made multiple tc= capabilities in
  a termcap entry work; recognize tc= also if it is not the last capability
  in an entry (Bugreport by Andrew Minter).

Release 9/3/03
* The code did not check st_blksize from stat(2) at other points besides
  the one fixed in the last release.
* The keyboard input character with code 255 ("y in ISO-8859-1) was
  misinterpreted as end-of-file indicator.

Release 8/27/03
* Compile fixes for AIX and HP-UX (Mike Jetzer).
* Delete temporary file when preserving was successful after receiving
  SIGHUP or SIGTERM (Fix taken from 4.3BSD Reno).
* Set MAILRC to /dev/null in expreserve to avoid reading the user's
  mail initialization file.
* Optionally use Caldera's 'UNIX(R) Regular Expression Library' to
  get POSIX.2 compatible REs.
* Don't refuse to quit with 'No write since last change' if a line of a
  newly read input file is too long but no changes were made by the user.
* The POSIX_2 preprocessor define has been removed. The behavior previously
  enabled by this variable is now the default (except as noted below).
* Backslash inside RE bracket expresssions (as in [\]]) is now recognized
  as an escape character by default. This is not POSIX.2 compliant, but is
  compatible with historic vi behavior, is demanded by the some traditional
  standards such as the System V Interface definition, and is compatible
  with vim. To get POSIX.2 compliant behavior here, define NO_BE_BACKSLASH.
* The input buffer did overflow with large values of st_blksize from stat(2).

Release 4/3/02
* Option "showmode" has no effect on hardcopy or glass terminals (P).
* Fixed undo for :r! command.
* Interrupt character is not misinterpreted as DEL on insertion (P).
* Fixed interaction of <count>~ vi command with abbreviations and macros (P).
* Avoid horizontal scrolling of showmode string on wraparound to last line (P).
* No showmode string is printed when executing a macro.
* Recovery listing fixed not to mix characters from long filenames in
  /var/preserve with /var/tmp listing (P).
* Catch SIGXFSZ (file size limit exceeded).

Release 3/3/02
* Separated terse/noterse messages for RE errors (P).
* The expreserve and exrecover commands work again (P).
* Passing beyond EOL in <count>~ vi command is not an error (P).
* Fixed segmentation violation on mdjc'd and related bugs (Sven Mascheck).
* Marks remain on lines changed in an undo operation.
* Close mapmalloc file descriptor for /dev/zero on exec() (P).
* Added -L and -V command line options as in SVr4 ex.
* POSIX.2: Processing EXINIT does not depend on stdin being a terminal.
* POSIX.2: No newline is appended to an empty file in visual mode.
* Fixed segmentation violation on :f^V^M^M and similar commands.
* Mapmalloc extended to allocate and release multiple memory pools.
* Exrecover lists /var/tmp in addition to /var/preserve.
* Have multiple attempts to create named buffer temporary file.
* Size limit for temporary file names removed.

Release 2/17/02

heirloom/diff/diff.c
/*
 * This code contains changes by
 * Gunnar Ritter, Freiburg i. Br., Germany, March 2003. All rights reserved.
 *
 * Conditions 1, 2, and 4 and the no-warranty notice below apply
 * to these changes.
 *
 *
 * Copyright (c) 1991
 * 	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 * 	This product includes software developed by the University of
 * 	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *
 * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *   Redistributions of source code and documentation must retain the
 *    above copyright notice, this list of conditions and the following
 *    disclaimer.
 *   Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *   All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed or owned by Caldera
 *      International, Inc.
 *   Neither the name of Caldera International, Inc. nor the names of
 *    other contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
 * INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
 * LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*	Sccsid @(#)diff.c	1.21 (gritter) 12/5/04>	*/
/*	from 4.3BSD diff.c 4.6 4/3/86	*/

#include "diff.h"
#include <unistd.h>
#include <locale.h>
/*
 * diff - driver and subroutines
 */

const char	diff[] = "diff";
const char	diffh[] = DIFFH;
const char	pr[] = "pr";
const char	*progname;
const char	*argv0;

static void	usage(void);
static void	xadd(const char *);

int
main(int argc, char **argv)
{
	int	i, invalid = 0;

	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	if (getenv("SYSV3") != NULL)
		sysv3 = 1;
	ifdef1 = "FILE1"; ifdef2 = "FILE2";
	status = 2;
	argv0 = argv[0];
	diffargv = argv;
	while ((i = getopt(argc, argv, ":D:efnbBwitcC:uU:hS:rslNx:a12"))
			!= EOF) {
		switch (i) {
#ifdef notdef
		case 'I':
			opt = D_IFDEF;
			wantelses = 0;
			break;
		case 'E':
			opt = D_IFDEF;
			wantelses = 1;
			break;
		case '1':
			opt = D_IFDEF;
			ifdef1 = optarg;
			break;
#endif
		case 'D':
			/* -Dfoo = -E -1 -2foo */
			wantelses = 1;
			ifdef1 = "";
			/* fall through */
#ifdef notdef
		case '2':
#endif
			opt = D_IFDEF;
			ifdef2 = optarg;
			break;
		case 'e':
			opt = D_EDIT;
			break;
		case 'f':
			opt = D_REVERSE;
			break;
		case 'n':
			opt = D_NREVERSE;
			break;
		case 'b':
			bflag = 1;
			break;
		case 'B':
			Bflag = 1;
			break;
		case 'w':
			wflag = 1;
			break;
		case 'i':
			iflag = 1;
			break;
		case 't':
			tflag = 1;
			break;
		case 'c':
			opt = D_CONTEXT;
			context = 3;
			break;
		case 'C': {
			char	*x;

			opt = D_CONTEXT;
			context = strtol(optarg, &x, 10);
			if (*x != '\0' || *optarg == '+' || *optarg == '-') {
				fprintf(stderr, "%s: use -C num\n", progname);
				done();
			}
			break;
		}
		case 'u':
			opt = D_UNIFIED;
			context = 3;
			break;
		case 'U': {
			char	*x;

			opt = D_UNIFIED;
			context = strtol(optarg, &x, 10);
			if (*x != '\0' || *optarg == '+' || *optarg == '-') {
				fprintf(stderr, "%s: use -U num\n", progname);
				done();
			}
			break;
		}
		case 'h':
			hflag++;
			break;
		case 'S':
			start = optarg;
			break;
		case 'r':
			rflag++;
			break;
		case 's':
			sflag++;
			break;
		case 'l':
			lflag++;
			break;
		case 'N':
			Nflag |= 3;
			break;
		case '1':
			Nflag |= 1;
			break;
		case '2':
			Nflag |= 2;
			break;
		case 'x':
			xadd(optarg);
			break;
		case 'a':
			aflag++;
			break;
		default:
			if (invalid == 0 && !sysv3)
				invalid = optopt;
		}
	}
	argv += optind, argc -= optind;
	if (argc != 2) {
		fprintf(stderr, sysv3 ? "%s: arg count\n" :
				"%s: two filename arguments required\n",
				progname);
		done();
	}
	file1 = argv[0];
	file2 = argv[1];
	if (invalid) {
		fprintf(stderr, "%s: invalid option -%c\n", progname, invalid);
		usage();
	}
	if (hflag && opt) {
		fprintf(stderr,
		    "%s: -h doesn't support -e, -f, -n, -c, -u, or -I\n",
		    	progname);
		done();
	}
	diffany(argv);
	/*NOTREACHED*/
	return 0;
}

void
diffany(char **argv)
{
	if (!strcmp(file1, "-"))
		stb1.st_mode = S_IFREG;
	else if (stat(file1, &stb1) < 0) {
		if (sysv3)
			stb1.st_mode = S_IFREG;
		else {
			fprintf(stderr, "%s: %s: %s\n", progname, file1,
					strerror(errno));
			done();
		}
	}
	if (!strcmp(file2, "-"))
		stb2.st_mode = S_IFREG;
	else if (stat(file2, &stb2) < 0) {
		if (sysv3)
			stb2.st_mode = S_IFREG;
		else {
			fprintf(stderr, "%s: %s: %s\n", progname, file2,
					strerror(errno));
			done();
		}
	}
	if ((stb1.st_mode & S_IFMT) == S_IFDIR &&
	    (stb2.st_mode & S_IFMT) == S_IFDIR) {
		diffdir(argv);
	} else
		diffreg();
	done();
}

static void
usage(void)
{
	fprintf(stderr, "\
usage: %s [ -bcefhilnrstw -C num -D string -S name ] file1 file2\n",
		progname);
	done();
}

int
min(int a,int b)
{

	return (a < b ? a : b);
}

int
max(int a,int b)
{

	return (a > b ? a : b);
}

void
done(void)
{
	if (tempfile1) {
		unlink(tempfile1);
		tempfile1 = NULL;
	}
	if (tempfile2) {
		unlink(tempfile2);
		tempfile2 = NULL;
	}
	if (recdepth == 0)
		exit(status);
	else
		longjmp(recenv, 1);
}

static void	noroom(void);

void *
dalloc(size_t n)
{
	struct stackblk *sp;

	if ((sp = malloc(n + sizeof *sp)) != NULL) {
		sp->s_prev = NULL;
		sp->s_next = curstack;
		if (curstack)
			curstack->s_prev = sp;
		curstack = sp;
		return (char *)sp + sizeof *sp;
	} else
		return NULL;
}

void *
talloc(size_t n)
{
	register void *p;

	if ((p = dalloc(n)) == NULL)
		noroom();
	return p;
}

void *
ralloc(void *p,size_t n)
{
	struct stackblk	*sp, *sq;

	if (p == NULL)
		return talloc(n);
	sp = (struct stackblk *)((char *)p - sizeof *sp);
	if ((sq = realloc(sp, n + sizeof *sp)) == NULL)
		noroom();
	if (sq->s_prev)
		sq->s_prev->s_next = sq;
	if (sq->s_next)
		sq->s_next->s_prev = sq;
	if (curstack == sp)
		curstack = sq;
	return (char *)sq + sizeof *sq;
}

void
tfree(void *p)
{
	struct stackblk	*sp;

	if (p == NULL)
		return;
	sp = (struct stackblk *)((char *)p - sizeof *sp);
	if (sp->s_prev)
		sp->s_prev->s_next = sp->s_next;
	if (sp->s_next)
		sp->s_next->s_prev = sp->s_prev;
	if (sp == curstack)
		curstack = sp->s_next;
	free(sp);
}

void
purgestack(void)
{
	struct stackblk *sp = curstack, *sq = NULL;

	do {
		free(sq);
		sq = sp;
		if (sp)
			sp = sp->s_next;
	} while (sq);
}

static void
noroom(void)
{
	oomsg(": files too big, try -h\n");
	status = 2;
	done();
}

static void
xadd(const char *cp)
{
	struct xclusion	*xp;

	xp = talloc(sizeof *xp);
	xp->x_pat = cp;
	xp->x_nxt = xflag;
	xflag = xp;
}

void
oomsg(const char *s)
{
	write(2, progname, strlen(progname));
	write(2, s, strlen(s));
}
